widget/entry.go

```diff
@@ -297,12 +486,122 @@ func (e *Entry) TypedRune(r rune) {
 	Renderer(e).(*entryRenderer).moveCursor()
 }
 
+// KeyDown handler for keypress events - used to store shift modifier state for text selection
+func (e *Entry) KeyDown(key *fyne.KeyEvent) {
+	// For keyboard cursor controlled selection we now need to store shift key state and selection "start"
+	// Note: selection start is where the highlight started (if the user moves the selection up or left then
+	// the selectRow/Column will not match SelectionStart)
+	if key.Name == desktop.KeyShiftLeft || key.Name == desktop.KeyShiftRight {
+		if e.selecting == false {
+			e.selectRow = e.CursorRow
+			e.selectColumn = e.CursorColumn
+		}
+		e.selectKeyDown = true
+	}
+}
+
+// KeyUp handler for key release events - used to reset shift modifier state for text selection
+func (e *Entry) KeyUp(key *fyne.KeyEvent) {
+	// Handle shift release for keyboard selection
+	// Note: if shift is released then the user may repress it without moving to adjust their old selection
+	if key.Name == desktop.KeyShiftLeft || key.Name == desktop.KeyShiftRight {
+		e.selectKeyDown = false
+	} else {
+		if e.selectKeyDown == false {
+			e.selecting = false
+		}
+	}
+}
+
+// eraseSelection removes the current selected region and moves the cursor
+func (e *Entry) eraseSelection() {
+	if e.ReadOnly {
+		return
+	}
+
+	provider := e.textProvider()
+	posA := e.SelectionStart()
+	posB := e.SelectionEnd()
+
+	if posA == posB {
+		return
+	}
+
+	e.Lock()
+
+	provider.deleteFromTo(posA, posB)
+	e.CursorRow, e.CursorColumn = e.rowColFromTextPos(posA)
+	e.selectRow, e.selectColumn = e.CursorRow, e.CursorColumn
+	e.Unlock()
+	e.selecting = false
+}
+
 // TypedKey receives key input events when the Entry widget is focused.
 func (e *Entry) TypedKey(key *fyne.KeyEvent) {
 	if e.ReadOnly {
 		return
 	}
+
+	// seeks to the start/end of the selection - used by: up, down, left, right
+	seekSelection := func(start bool) {
+		var pos int
+		if start {
+			pos = e.SelectionStart()
+		} else {
+			pos = e.SelectionEnd()
+		}
+		e.Lock()
+		e.CursorRow, e.CursorColumn = e.rowColFromTextPos(pos)
+		e.Unlock()
+		e.selecting = false
+	}
+
 	provider := e.textProvider()
+
+	if e.selectKeyDown && e.selecting == false {
+		switch key.Name {
+		case fyne.KeyUp, fyne.KeyDown,
+			fyne.KeyLeft, fyne.KeyRight,
+			fyne.KeyEnd, fyne.KeyHome,
+			fyne.KeyPageUp, fyne.KeyPageDown:
+			e.selecting = true
+		}
+	}
+
+	if e.selecting {
+
+		processKey := true
+		switch key.Name {
+		case fyne.KeyBackspace, fyne.KeyDelete:
+			// clears the selection
+			e.eraseSelection()
+			processKey = false
+		case fyne.KeyReturn, fyne.KeyEnter:
+			e.eraseSelection() // clear the selection and fallthrough to add the newline
+		}
+
+		if e.selectKeyDown == false {
```

Comment:
It is my understanding of this code that this should always be true whilst e.selecting is true. 
I am struggling to understand the following code section, and I wonder if there is a possibility it is not executed.
If this code is required, along with the seekSelection above, I think it should be restructured or somehow made easier to read.

Feel free to join our Slack conversations if you'd like to discuss in person :)

Reply:
selectKeyDown will be false when selecting is true in the case where a selection is visible but the user is no longer holding left or right shift keys.
I've grouped similar keypresses (left and right cursor, up and down cursor, etc) into individual cases in the select as a means to simplify the logic as much as possible. A complication that I've tried to make as visible as possible is are the conditions under which the key should be processed as normal by the original select statement which immediately follows the outer `if e.selecting {` scope.

Reply:
I wonder if these up/down left/right special cases are perhaps not needed, in all these situations I would have thought that the selection ends and the key is processed.
For example if I have a selection and then press up I would expect to cancel the selection and then go to the previous line directly above the current cursor?
I suspect you have included some functionality here that provides additional behaviour?